app/models/pipeline_run.rb

```diff
@@ -153,38 +162,317 @@ def retry
+    self.fraction_subsampled = subsample_fraction
+    save
+
+    version_s3_path = "#{alignment_output_s3_path}/#{VERSION_JSON_NAME}"
+    update(version: `aws s3 cp #{version_s3_path} -`)
+
+    # only keep counts at certain taxonomic levels
+    taxon_counts_attributes_filtered = []
+    acceptable_tax_levels = [TaxonCount::TAX_LEVEL_SPECIES]
+    acceptable_tax_levels << TaxonCount::TAX_LEVEL_GENUS if multihit?
+    acceptable_tax_levels << TaxonCount::TAX_LEVEL_FAMILY if multihit?
+    pipeline_output_dict['taxon_counts_attributes'].each do |tcnt|
+      # TODO:  Better family support.
+      if acceptable_tax_levels.include?(tcnt['tax_level'].to_i) && !invalid_family_call?(tcnt)
+        taxon_counts_attributes_filtered << tcnt
+      end
+    end
+    # Set created_at and updated_at
+    current_time = Time.now.utc # to match TaxonLineage date range comparison
+    taxon_counts_attributes_filtered.each do |tcnt|
+      tcnt["created_at"] = current_time
+      tcnt["updated_at"] = current_time
+    end
+    update(taxon_counts_attributes: taxon_counts_attributes_filtered)
+
+    # aggregate the data at genus level
+    generate_aggregate_counts('genus') unless multihit?
+    # merge more accurate name information from lineages table
+    update_names
+    # denormalize superkingdom_taxid into taxon_counts
+    if multihit?
+      update_superkingdoms
+    else
+      update_genera
+    end
+    # label taxa as phage or non-phage
+    update_is_phage
+
+    # rm the json
+    _stdout, _stderr, _status = Open3.capture3("rm -f #{downloaded_json_path}")
+  end
+
+  def db_load_byteranges
+    byteranges_json_s3_path = "#{postprocess_output_s3_path}/#{TAXID_BYTERANGE_JSON_NAME}"
+    downloaded_byteranges_path = PipelineRun.download_file(byteranges_json_s3_path, local_json_path)
+    taxon_byteranges_csv_file = "#{local_json_path}/taxon_byteranges"
+    hash_array_json2csv(downloaded_byteranges_path, taxon_byteranges_csv_file, %w[taxid hit_type first_byte last_byte])
+    ` cd #{local_json_path};
+      sed -e 's/$/,#{id}/' -i taxon_byteranges;
+      mysqlimport --replace --local --user=$DB_USERNAME --host=#{rds_host} --password=$DB_PASSWORD --columns=taxid,hit_type,first_byte,last_byte,pipeline_run_id --fields-terminated-by=',' idseq_#{Rails.env} taxon_byteranges;
+    `
+    _stdout, _stderr, _status = Open3.capture3("rm -f #{downloaded_byteranges_path}")
+  end
+
+  def s3_file_for(output)
+    case output
+    when "ercc_counts"
+      "#{host_filter_output_s3_path}/#{ERCC_OUTPUT_NAME}"
+    when "taxon_counts"
+      "#{alignment_output_s3_path}/#{taxon_counts_json_name}"
+    when "taxon_byteranges"
+      "#{postprocess_output_s3_path}/#{TAXID_BYTERANGE_JSON_NAME}"
+    end
+  end
+
+  def output_ready?(output)
+    file_generated_since_run(s3_file_for(output))
+  end
+
+  def update_result_status(field, value)
+    status_hash = result_status_hash
+    status_hash[field] = value
+    update(result_status: status_hash.to_json)
+  end
```

Comment:
How do you ensure atomicity for these updates?

Reply:
fixed according to our discussion